### ISSUE TITLE
Proposed refactoring of Bot, Updater and Dispatcher

### ISSUE THREAD

**Issue by rbpasker** [python-telegram-bot/python-telegram-bot#116]
Having worked with the API for full day now, I feel fully qualified (ahem) to propose the following refactoring:

1. rename `Bot` to `Telegram`, since its real purpose is not necessarily as a bot, but rather as a Python wrapper for the Telegram web API.

2. Remove the threading from `Dispatcher` to make it a pure dispatcher, which will allow the Handler functionality, and the proposed decorators seen in PR #47, to be used by both threaded and unthreaded bots.

3. Add decorators to 'Dispatcher' for each type of handler, like `@command('start')` and `@message` and `@regexMessage(".*")` h/t to PR #47.

4. Rename `Updater` to `ThreadedDispatcher`, and move all the threading code from the current `Dispatcher` into it.  

5. Add a `WebhookDispatcher` as a subclass of `ThreadedDispatcher` to operate via webhooks.

6. Add `GenericBot` and `GenericThreadedBot` classes which actual bots can subcalss and have functionality to manage tokens externally (eg, in a file or on the command line), perform the `getMe()`,  initialize logging, handle `/start` and `/help` (using docstrings).  Make examples subclass these generic bots.  Optional: My bots also have `//` commands to control the bot during debugging, such as `//verbose` and `//logging`, and we could add these to the generic bots.

I realize that this is a fairly substantial change, but I think its conceptually closer to what we want.

**Comment by jh0ker** [comment]
1. Yes, that seems reasonable. 

2. There is currently no threading code in Dispatcher except for the @run_async decorator and related, which is optional to use. 

3. Again, that seems reasonable 

4. I'm not sure how that will improve the experience, also see 2. The Updaters only purpose is to get updates from telegram and to start the Dispatcher in a separate thread. We could move that functionality to a new class, however. 

5. Discussable, but depends on 2 and 4

6. That seems nice as well

**Comment by rbpasker** [comment]
I see where the confusion is in # 2, so let me rephrase:

2. So that `Dispatcher` can be used by threaded and non-threaded dispatcher,  it should be modified so that it no longer operates on a queue, but rather operates by calling the existing `process_update()` or by calling a new method `process_udpates()` that takes an array of updates.  Move all queue operations into `ThreadedDispatcher`, which can then call `Dispatcher::process_update[s]()`.

**Comment by jh0ker** [comment]
So, the `Dispatcher` right now can be used with or without threading already. Just call `processUpdate()` on it. `start` and `stop` methods are optional to use, basically. We could add an option in the `Updater` not to run it in a Thread but to directly call `processUpdate()`, but why? What is the improvement?

Decorators we can have either way, having threads doesn't change that. 

Renaming `Updater` into `ThreadedDispatcher`? But the updater doesn't dispatch anything...

**Comment by rahiel** [comment]
> 1. rename Bot to Telegram, since its real purpose is not necessarily as a bot, but rather as a Python wrapper for the Telegram web API.

I disagree with this proposal, this library is a wrapper around Telegram's [Bot API](https://core.telegram.org/bots/api). All the methods work on a specific `Bot` instance. Renaming to `Telegram` is confusing: while there is only one Telegram, there are many different bots. (You can have several `Bot` objects in a program, but there is only one Telegram.)

This would also break existing code!
